packages/editor/src/components/post-text-editor/index.js

```diff
@@ -65,6 +66,14 @@ export default function PostTextEditor() {
 		}
 	};
 
+	// Ensure changes aren't lost when switching modes using shortcuts.
+	useShortcut( 'core/edit-post/toggle-mode', () => {
+		if ( isDirty ) {
+			const blocks = parse( value );
+			resetEditorBlocks( blocks );
+		}
+	} );
```

Comment:
This feels bit hacky, but works better than plain side-effect.

@youknowriad, @ntsekouras I would love to get your opinion on this method.